src/com/amazon/corretto/crypto/provider/Loader.java

```diff
@@ -289,20 +324,78 @@ private static void assertVersionMatch() {
         }
     }
 
+    private static void maybeDelete(final Path path) {
+        if (!DebugFlag.PRESERVE_NATIVE_LIBRARIES.isEnabled()) {
+            try {
+                Files.delete(path);
+            } catch (IOException ex) {
+                LOG.warning("Unable to delete native library: " + ex);
+            }
+        }
+    }
+
+    private static void maybeDeletePrivateTempDir(final Path tmpDirectory) {
+        for (String jarResource: JAR_RESOURCES) {
+            String resourceFileName = System.mapLibraryName(jarResource);
+            maybeDelete(tmpDirectory.resolve(resourceFileName));
+        }
+        maybeDelete(tmpDirectory);
+    }
+
+    private static Path writeJarResourceToTemporaryFile(Path tempDirectory, final String resourceFileName) throws IOException {
+        final Path tempResourceFilePath = tempDirectory.resolve(resourceFileName);
+
+        try (InputStream is = Loader.class.getResourceAsStream(resourceFileName);
+             OutputStream os = Files.newOutputStream(tempResourceFilePath, StandardOpenOption.CREATE,
+                     StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING)) {
```

Comment:
We need to ensure that file permissions are properly set at creation time (like with the temporary directory).

We also need to fail if an a file already exists with this name and not overwrite it.

Reply:
Fixed